contracts/ERC20Decrypto.sol

```diff
@@ -210,11 +231,257 @@ contract ERC20Decrypto is
+     * `subtractedValue`.
+     */
+    function decreaseAllowance(address spender, uint256 subtractedValue)
+        public
+        virtual
+        override
+        returns (bool)
+    {
+        // uint256 splitValue = _underlyingValue(subtractedValue);
+        uint256 formattedValue =
+            _formattedValue(_allowances[_msgSender()][spender]);
+
+        _approve(
+            _msgSender(),
+            spender,
+            formattedValue.sub(
+                subtractedValue,
+                "ERC20: decreased allowance below zero"
+            )
+        );
+        return true;
+    }
+
+    /**
+     * @dev See {IERC20-totalSupply}.
+     */
+    function totalSupply() public view virtual override returns (uint256) {
+        return _formattedValue(_totalSupply);
+    }
+
+    /**
+     * @dev Destroys `amount` tokens from the caller.
+     *
+     * See {ERC20-_burn}.
+     */
+    function burn(uint256 amount) public virtual override {
+        uint256 splitAmount = _underlyingValue(amount);
+        _burn(_msgSender(), splitAmount);
+    }
+
+    /**
+     * @dev Destroys `amount` tokens from `account`, deducting from the caller's
+     * allowance.
+     *
+     * See {ERC20-_burn} and {ERC20-allowance}.
+     *
+     * Requirements:
+     *
+     * - the caller must have allowance for ``accounts``'s tokens of at least
+     * `amount`.
+     */
+    function burnFrom(address account, uint256 amount) public virtual override {
+        uint256 udnerlyingAmount = _underlyingValue(amount);
+        uint256 decreasedAllowance =
+            allowance(account, _msgSender()).sub(
+                amount,
+                "ERC20: burn amount exceeds allowance"
+            );
+        //TODO send approv with out underlying
+        _approve(account, _msgSender(), decreasedAllowance);
+        _burn(account, udnerlyingAmount);
+    }
+
+    /**
+     * @dev See {IERC20-transferFrom}.
+     *
+     * Emits an {Approval} event indicating the updated allowance. This is not
+     * required by the EIP. See the note at the beginning of {ERC20}.
+     *
+     * Requirements:
+     *
+     * - `sender` and `recipient` cannot be the zero address.
+     * - `sender` must have a balance of at least `amount`.
+     * - the caller must have allowance for ``sender``'s tokens of at least
+     * `amount`.
+     */
+    function transferFrom(
+        address sender,
+        address recipient,
+        uint256 amount
+    ) public virtual override returns (bool) {
+        _transfer(sender, recipient, amount);
+        uint256 underlyingAmount = _underlyingValue(amount);
+        uint256 formattedAmount =
+            _formattedValue(
+                _allowances[sender][_msgSender()].sub(
+                    underlyingAmount,
+                    "ERC20: transfer amount exceeds allowance"
+                )
+            );
+        _approve(sender, _msgSender(), formattedAmount);
```

Comment:
was this here before?

Reply:
because is not required, but emits approval indicating the updated allowance.
https://github.com/Think-and-Dev/decrypto/blob/main/contracts/zeppelin/token/ERC20/ERC20Upgradeable.sol#L158